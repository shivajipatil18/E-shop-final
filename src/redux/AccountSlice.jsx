import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';


export const fetchUser = createAsyncThunk('account/fetchUser', async (id, { rejectWithValue }) => {
  try {
    const res = await axios.get(`http://localhost:3001/users/${id}`);
    return res.data;
  } catch (err) {
    return rejectWithValue(err.response?.data || 'Failed to fetch user');
  }
});

export const updateUserInfo = createAsyncThunk(
  'account/updateUserInfo',
  async ({ id, data }, { rejectWithValue }) => {
    try {
      const res = await axios.patch(`http://localhost:3001/users/${id}`, data);
      return res.data;
    } catch (err) {
      return rejectWithValue(err.response?.data || 'Failed to update info');
    }
  }
);

// Change password
export const changePassword = createAsyncThunk(
  'account/changePassword',
  async ({ id, currentPassword, newPassword }, { rejectWithValue, getState }) => {
    try {
      const { account } = getState();
      if (account.user.password !== currentPassword) {
        return rejectWithValue('Current password is incorrect.');
      }
      const res = await axios.patch(`http://localhost:3001/users/${id}`, { password: newPassword });
      return res.data;
    } catch (err) {
      return rejectWithValue(err.response?.data || 'Password update failed');
    }
  }
);

const accountSlice = createSlice({
  name: 'account',
  initialState: {
    user: null,
    status: 'idle',
    error: null,
  },
  reducers: {
    clearAccountState: (state) => {
      state.user = null;
      state.status = 'idle';
      state.error = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchUser.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchUser.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.user = action.payload;
      })
      .addCase(fetchUser.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload;
      })
      .addCase(updateUserInfo.fulfilled, (state, action) => {
        state.user = { ...state.user, ...action.payload };
      }).addCase(updateUserInfo.rejected, (state, action) => {
        state.error = action.payload;
      })
      .addCase(changePassword.fulfilled, (state, action) => {
        state.user = { ...state.user, password: action.payload.password };
      }).addCase(changePassword.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || action.error.message;
      });;
  },
});

export const { clearAccountState } = accountSlice.actions;
export default accountSlice.reducer;
