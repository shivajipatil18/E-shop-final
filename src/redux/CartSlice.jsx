import { createSlice } from "@reduxjs/toolkit";

let initialState = {
  products: [],
  totalQuantity: 0,
  totalPrice: 0,
  searchResults: [],
};
let storedCart ="";
if (typeof window !== "undefined") {
storedCart = window.localStorage.getItem("cart");
}
if (storedCart) {
  initialState = JSON.parse(storedCart);
}
const cartSlice = createSlice({
  name: "product",
  initialState,
  reducers: {
    searchProduct: (state, action) => {
      const query = action.payload.toLowerCase();
      state.searchResults = state.products.filter((product) =>
        product.title.toLowerCase().includes(query)
      );
    },
  
    addToCart(state, action) {
      const newItem = action.payload;
      const itemIndex = state.products.findIndex(
        (item) => item.id === newItem.id
      );
    
      if (itemIndex !== -1) {
        state.products[itemIndex].quantity++;
        state.products[itemIndex].totalPrice += newItem.price;
      } else {
        state.products.push({
          id: newItem.id,
          title: newItem.title,
          price: newItem.price,
          quantity: 1,
          totalPrice: newItem.price,
          image: newItem.image,
        });
      }
    
      state.totalQuantity = state.products.reduce((sum, item) => sum + item.quantity, 0);
      state.totalPrice = state.products.reduce((sum, item) => sum + item.quantity * item.price, 0)
    },
    
    removeFromCart(state, action) {
      const id = action.payload;
      const findItem = state.products.findIndex((item) => item.id === id);
      if (findItem !== -1) {
        state.totalPrice -= state.products[findItem].totalPrice;
        state.totalQuantity -= state.products[findItem].quantity;
        state.products = state.products.filter((item) => item.id !== id);
      }
    },
    increaseQuantity(state, action) {
      const id = action.payload;
      const findItem = state.products.findIndex((item) => item.id === id);
      if (findItem !== -1) {
        state.products[findItem].quantity++;
        state.products[findItem].totalPrice += state.products[findItem].price;
        state.totalQuantity++;
        state.totalPrice += state.products[findItem].price;
      }
    },
    decreaseQuantity(state, action) {
      const id = action.payload;
      const findItem = state.products.findIndex((item) => item.id === id);
      if (findItem !== -1 && state.products[findItem].quantity > 1) {
        state.products[findItem].quantity--;
        state.products[findItem].totalPrice -= state.products[findItem].price;
        state.totalQuantity--;
        state.totalPrice -= state.products[findItem].price;
      }
      
    },
    setCart: (state, action) => {
      return action.payload;
    },
    clearCart(state) {
      state.products = [];
    },
  },  
  
});
export const { addToCart, removeFromCart, increaseQuantity, decreaseQuantity,setCart,searchProduct,clearCart } =
  cartSlice.actions;
export default cartSlice.reducer;
