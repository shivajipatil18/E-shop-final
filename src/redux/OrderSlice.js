
import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios, { all } from "axios";

 export const fetchAllOrders = createAsyncThunk('orders/fetchOrders', async () => {
   const response = await axios.get('http://localhost:5000/orders');   
    return response.data;

 });

export const saveOrder = createAsyncThunk("order/saveOrder", async (order, thunkAPI) => {
  try {
    const response = await axios.post("http://localhost:5000/orders", order);
      return response.data;
    
  } catch (error) {
    return thunkAPI.rejectWithValue(error.message);
  }
});
  export const fetchOrders = createAsyncThunk("order/fetchOrders", async (username, thunkAPI) => {
    try {
      const response = await axios.get(`http://localhost:5000/orders?username=${username}`); 
      return response.data;
    } catch (error) {
      return thunkAPI.rejectWithValue(error.message);
    }
  });
  export const updateOrderStatus = createAsyncThunk(
    "order/updateOrderStatus",
    async ({ orderId, newStatus }) => {
      const response = await fetch(`http://localhost:5000/orders/${orderId}/status`, { 
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({ status: newStatus }),
      });
  console.log(response,orderId,newStatus);
      if (!response.ok) {
        throw new Error("Failed to update order status");
      }
  
      return await response.json(); 
    }
  );
  
const orderSlice = createSlice({
  name: "order",
  initialState: {
    orderDetails: [],
    loading: false,
    error: null,
    allOrders: [],
  },
  reducers: {
    setOrderDetails: (state, action) => {
      state.orderDetails = [action.payload];
    },
    clearOrderDetails: (state) => {
      state.orderDetails = [];
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(saveOrder.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(saveOrder.fulfilled, (state, action) => {
        state.loading = false;
        state.orderDetails = [action.payload];
      })
      .addCase(saveOrder.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      }) .addCase(fetchOrders.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchOrders.fulfilled, (state, action) => {
        state.loading = false;
        state.orderDetails = action.payload;
      })
      .addCase(fetchOrders.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      })  .addCase(fetchAllOrders.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchAllOrders.fulfilled, (state, action) => {
        state.loading = false;
        state.allOrders = action.payload;
      })
      // .addCase(fetchAllOrders.rejected, (state, action) => {
      //   state.loading = false;
      //   state.error = action.payload || "Something went wrong";
      // })  .addCase(updateOrderStatus.pending, (state) => {
      //   state.loading = true;
      //   state.error = null;
      // })
      // .addCase(updateOrderStatus.fulfilled, (state, action) => {
      //   state.loading = false;
      //   const updatedOrder = action.payload;
      //   const index = state.allOrders.findIndex((order) => order.id === updatedOrder.orderId);
      //   if (index !== -1) {
      //     state.allOrders[index].status = updatedOrder.newStatus;
      //   }
      // })
      // .addCase(updateOrderStatus.rejected, (state, action) => {
      //   state.loading = false;
      //   state.error = action.payload || action.error.message;
      // });
  },
});
      
  


export const { setOrderDetails, clearOrderDetails ,} = orderSlice.actions;
export default orderSlice.reducer;
