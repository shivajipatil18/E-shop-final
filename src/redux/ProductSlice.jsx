import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

// Fetch all products
export const fetchProducts = createAsyncThunk('products/fetchProducts', async () => {
  const response = await axios.get('https://fakestoreapi.com/products');
  return response.data;
});

// Add a new product
export const addProduct = createAsyncThunk('products/addProduct', async (product) => {
  const response = await axios.post('https://fakestoreapi.com/products', product);
  return {
    ...response.data,
    rating: { rate: 0, count: 0 }, 
  };
});


export const updateProduct = createAsyncThunk('products/updateProduct', async (product) => {
  const response = await axios.put(`https://fakestoreapi.com/products/${product.id}`, product);
  return response.data;
});


export const deleteProduct = createAsyncThunk('products/deleteProduct', async (id) => {
  await axios.delete(`https://fakestoreapi.com/products/${id}`);
  return id;
});


export const selectFilteredProducts = (state) => {
  const { products = [], category, price } = state.product || {};
  return products.filter((product) => {
    const matchCategory = category ? product.category === category : true;
    const matchPrice = product.price <= price;
    // const matchRating = typeof product.rating === 'number' ? product.rating >= minRating : false;

    return matchCategory && matchPrice //&& matchRating;
    
  });
};

const initialState = {
  products: [],
  searchTerm: "",
  filteredData: [],
  loading: false,
  error: null,
  category: "",
  price: 1000,
  // minRating: 4, 

};

const productSlice = createSlice({
  name: "product",
  initialState,
  reducers: {
    setCategory: (state, action) => {
      state.category = action.payload;
    },
    setPrice: (state, action) => {
      state.price = action.payload;
    },
    setSearchTerm: (state, action) => {
      state.searchTerm = action.payload;
    },
    setFilteredData: (state, action) => {
      state.filteredData = action.payload;
    },
    setFilteredProducts(state, action) {
      state.filteredProducts = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(fetchProducts.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchProducts.fulfilled, (state, action) => {
        state.loading = false;
        state.products = [...state.products, ...action.payload];
        state.filteredData = action.payload;
      })
      .addCase(fetchProducts.rejected, (state, action) => {
        state.loading = false;
        state.error = action.error.message;
      })
  
  },
});

export const { setCategory, setPrice, setSearchTerm, setFilteredData,setFilteredProducts } = productSlice.actions;
export default productSlice.reducer;