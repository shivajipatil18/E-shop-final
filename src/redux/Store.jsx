import { configureStore } from "@reduxjs/toolkit";

import productReducer from "./ProductSlice";
import cartReducer from "./CartSlice";
import authReducer from './authSlice';
import orderReducer from "./OrderSlice";
import accountReducer from "./AccountSlice"
import userReducer from "./UserSlice";
import { use } from "react";

const store = configureStore({
  reducer: {
    product: productReducer,
    cart: cartReducer,
    auth: authReducer,
    order: orderReducer,
    account: accountReducer,
    user: userReducer,
  },
});
export default store;