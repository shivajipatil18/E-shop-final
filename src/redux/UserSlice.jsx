import { createSlice, createAsyncThunk } from "@reduxjs/toolkit";
import axios from "axios";

export const fetchAllUsers = createAsyncThunk("user/fetchAllUsers", async (_, thunkAPI) => {
  try {
    const response = await axios.get("http://localhost:5000/users");
    return response.data;
  } catch (error) {
    return thunkAPI.rejectWithValue(error.message);
  }
});

const userSlice = createSlice({
  name: "user",
  initialState: {
    users: [],
    loading: false,
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchAllUsers.pending, (state) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchAllUsers.fulfilled, (state, action) => {
        state.loading = false;
        state.users = action.payload;
      })
      .addCase(fetchAllUsers.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      });
  },
});

export default userSlice.reducer;
