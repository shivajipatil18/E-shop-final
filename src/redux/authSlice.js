import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const register = createAsyncThunk('auth/register', async ({ name, email, password }, { rejectWithValue }) => {
  try {
    const response = await axios.post('http://localhost:3001/users', {
      name,
      email,
      password,
    });
    return response.data;
  } catch (error) {
    return rejectWithValue(error.response.data);
  }
});

export const login = createAsyncThunk('auth/login', async ({ email, password }, { rejectWithValue }) => {
  try {
    const response = await axios.get(`http://localhost:3001/users?email=${email}&password=${password}`);
     console.log(response);
     if(response.data.length && response.data[0].email !== email){
       return [];
     }
     return response.data;
  
  } catch (error) {
    return rejectWithValue(error.response.data);
  }
});

const authSlice = createSlice({
  name: 'auth',
  initialState: {
    isAuthenticated: false,
    status: 'idle',
    error: null,
    user: null,
  },
  reducers: {
    logout: (state) => {
      state.isAuthenticated = false;
      state.user = null;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(register.pending, (state) => {
        state.status = 'loading';
        state.error = null;
      })
      .addCase(register.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.isAuthenticated = true;
        state.user = action.payload;
        state.error = null;
      })
      .addCase(register.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || 'Registration failed';
      })
      .addCase(login.pending, (state) => {
        state.status = 'loading';
        state.error = null;
      })
      .addCase(login.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.isAuthenticated = true;
        state.user = action.payload;
        state.error = null;
      })
      .addCase(login.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.payload || 'Login failed';
      });
  },
});

export const { logout } = authSlice.actions;
export default authSlice.reducer;