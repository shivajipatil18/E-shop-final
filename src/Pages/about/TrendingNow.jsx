import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { selectFilteredProducts } from '../../redux/ProductSlice';
import { addToCart } from '../../redux/CartSlice';
import { toast } from 'react-toastify';

const TrendingNow = () => {
  const dispatch = useDispatch();
  const products = useSelector(selectFilteredProducts);

  const trendingProducts = (products || [])
    .filter((product) => product.rating?.rate > 4)
    .slice(0, 8);

  const handleAddToCart = (product) => {
    dispatch(addToCart(product));
    toast.success(`${product.title} has been added to your cart!`);

  };

  if (trendingProducts.length === 0) {
    return (
      <p className="text-center text-lg text-gray-500">
        No trending products available.
      </p>
    );
  }

  return (
    <section className="bg-gray-50 py-12">
      <div className="max-w-screen-xl mx-auto px-4">
        <h2 className="text-4xl font-extrabold text-center text-gray-800 mb-6">
          üî• Trending Now
        </h2>
        <p className="text-lg text-center text-gray-600 mb-12">
          Check out the hottest products currently trending in our store!
        </p>

        <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-8">
          {trendingProducts.map((product) => (
            <div
              key={product.id}
              className="bg-white rounded-2xl shadow-lg hover:shadow-2xl transition duration-300 transform hover:scale-105 overflow-hidden flex flex-col"
            >
              <img
                src={product.image || '/default-image.jpg'}
                alt={product.title || 'Product Image'}
                className="w-full h-60 object-contain bg-gray-100 p-4"
              />
              <div className="p-5 flex flex-col justify-between flex-grow">
                <h3 className="text-base font-semibold text-gray-800 line-clamp-2">
                  {product.title || 'Untitled'}
                </h3>
                <p className="text-sm text-gray-600 mt-1 line-clamp-2">
                  {product.description || ''}
                </p>
                <div className="mt-4 flex justify-between items-center">
                  <span className="text-lg font-bold text-red-600">
                    ‚Çπ{product.price !== undefined ? product.price : 'N/A'}
                  </span>
                  <button
                  data-testid="add-to-cart-btn"
                    onClick={() => handleAddToCart(product)}
                    className="bg-red-600 text-white px-4 py-1.5 rounded-full text-sm font-medium hover:bg-red-700 transition"
                  >
                    Add to Cart
                  </button>
                </div>
                <p className="text-sm text-gray-500 mt-2">
                  ‚≠ê {product.rating?.rate || 'N/A'} / 5
                </p>
              </div>
            </div>
          ))}
        </div>
      </div>
    </section>
  );
};

export default TrendingNow;
