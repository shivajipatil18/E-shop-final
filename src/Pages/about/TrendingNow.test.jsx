// TrendingNow.test.jsx (Enzyme version)
import React from "react";
import { shallow, mount } from "enzyme";
import TrendingNow from "./TrendingNow";
import * as redux from "react-redux";
import { addToCart } from "../../redux/CartSlice";

// Mock Redux hooks
jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe("TrendingNow Component with Enzyme", () => {
  const mockDispatch = jest.fn();

  const mockProducts = [
    { id: 1, title: "Product 1", rating: 4.5, price: 100 },
    { id: 2, title: "Product 2", rating: 3.5, price: 200 },
    { id: 3, title: "Product 3", rating: 5.0, price: 150 },
  ];

  beforeEach(() => {
    redux.useDispatch.mockReturnValue(mockDispatch);
    redux.useSelector.mockImplementation((selector) =>
      selector({
        product: {
          filteredProducts: mockProducts,
        },
      })
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render without crashing", () => {
    const wrapper = mount(<TrendingNow />);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.text()).toContain("Trending Now");
  });

  it("should show message when no trending products", () => {
    redux.useSelector.mockImplementation((selector) =>
      selector({
        product: {
          filteredProducts: [],
        },
      })
    );
    const wrapper = mount(<TrendingNow />);
    expect(wrapper.text()).toContain("No trending products");
  });

  it("should render only products with rating > 4", () => {
    const wrapper = mount(<TrendingNow />);
    expect(wrapper.text()).toContain("Product 1");
    expect(wrapper.text()).not.toContain("Product 2");
    expect(wrapper.text()).toContain("Product 3");
  });

  it("should dispatch addToCart action when Add to Cart button is clicked", () => {
    const wrapper = mount(<TrendingNow />);

    const addButtons = wrapper.find('[data-testid="add-to-cart-btn"]');

    expect(addButtons.length).toBeGreaterThan(0);

    addButtons.at(0).simulate("click");

    expect(mockDispatch).toHaveBeenCalled();
    expect(addToCart).toHaveBeenCalledWith(mockProducts[0]);
  });

  it("should match snapshot", () => {
    const wrapper = shallow(<TrendingNow />);
    expect(wrapper).toMatchSnapshot();
  });
});
