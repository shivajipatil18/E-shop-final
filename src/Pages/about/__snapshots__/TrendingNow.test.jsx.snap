// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrendingNow Component with Enzyme should match snapshot 1`] = `ShallowWrapper {}`;
